### PR TITLE
[AST] Improve ErrorType handling in TypeBase::getMemberSubstitutions()

### PR DESCRIPTION
**Pull request by jtbandes**
Historical note: a similar change was made in 0033bda.

Fixes 78 compiler_crashers. :boat:

### CODE REVIEW

**Comment by lattner** [comment]
This LGTM, but I think @DougGregor should also take a look when he gets a chance.  Great work, thank you!

-Chris

**Comment by AlexDenisov** [comment]
Indeed it solves some issues, but the problem actually caused by [this line](https://github.com/apple/swift/pull/272/files#diff-878271d251a2aeb73064c06c35030ae6R2520):

```
baseTy = baseTy->castTo<NominalType>()->getParent();
```

The best way to fix those crashes, imho, would be to use safer API for casting:

```
if (const NominalType *NominalTy = baseTy->getAs<NominalType>()) {
  baseTy = NominalTy->getParent();
} else {
  baseTy = nullptr;
}
```

otherwise some other non-`NominalType` and non-`ErrorType` will break the compiler.

**Comment by jtbandes** [comment]
@AlexDenisov thanks for the input. I would assume the `castTo<NominalType>` is being done for good reason, but I don't have enough knowledge to say for sure. @DougGregor will need to confirm.

Really makes me wish this were all written in Swift, so we could use optional chaining and `throws` :smile:

**Comment by jtbandes** [comment]
Updated with new fixed crashers.

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -2498,7 +2498,7 @@ TypeSubstitutionMap TypeBase::getMemberSubstitutions(DeclContext *dc) {
 
   // Gather all of the substitutions for all levels of generic arguments.
   GenericParamList *curGenericParams = dc->getGenericParamsOfContext();
-  while (baseTy) {
+  while (baseTy && !baseTy->is<ErrorType>()) {
```
Comment (DougGregor) [inline lib/AST/Type.cpp]: I'd rather that the ErrorType check happen a little later, e.g....

**Comment by jtbandes** [comment]
@DougGregor: Updated.

I have general questions about the error-handling mechanisms throughout; I sent an email to `swift-dev` entitled "Problems caused by is<ErrorType>", and I also filed https://bugs.swift.org/browse/SR-86. I'd be curious to hear your thoughts.

**Comment by jtbandes** [comment]
@DougGregor please let me know if there's anything else to be done here!

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -2517,7 +2517,17 @@ TypeSubstitutionMap TypeBase::getMemberSubstitutions(DeclContext *dc) {
     }
 
     // Continue looking into the parent.
-    baseTy = baseTy->castTo<NominalType>()->getParent();
+    auto nominalTy = baseTy->getAs<NominalType>();
+    if (!nominalTy || nominalTy->is<ErrorType>()) {
```
Comment (DougGregor) [inline lib/AST/Type.cpp]: The getAs will either produce a NominalType or null; there's no need to check for ErrorType here.

**Pull request merged**